File: Chapter08/Laser.cpp
//
//  Laser.cpp
//  Game-mac
//
//  Created by Michael Lehmann on 12.01.21.
//  Copyright © 2021 Michael Lehmann. All rights reserved.
//

#include "Laser.h"
#include "Game.h"
#include "Asteroid.h"
#include "SpriteComponent.h"
#include "MoveComponent.h"
#include "CircleComponent.h"

Laser::Laser(Game* game)
    :Actor(game)
    ,mDeathTimer(1.0f) {
    // Create a sprite component
    SpriteComponent* sc = new SpriteComponent(this);
    sc->SetTexture(game->GetTexture("Assets/Laser.png"));

    // Create a move component, and set a forward speed
    MoveComponent* mc = new MoveComponent(this);
    mc->SetForwardSpeed(800.0f);

    // Create a circle component (for collision)
    mCircle = new CircleComponent(this);
    mCircle->SetRadius(11.0f);
}

void Laser::UpdateActor(float deltaTime) {
    // If we run out of time, laser is dead
    mDeathTimer -= deltaTime;
    if (mDeathTimer <= 0.0f) {
        SetState(EDead);
    }
    else {
        // Do you intersect with an asteroid?
        for (auto ast : GetGame()->GetAsteroids()) {
            if (Intersect(*mCircle, *(ast->GetCircle()))) {
                // If this laser intersects with an asteroid,
                // set ourselves an the asteroid to dead
                SetState(EDead);
                ast->SetState(EDead);
                break;
            }
        }
    }
}

File: Chapter06_Veldrid/Laser.cs
namespace Chapter06
{
    public class Laser : Actor
    {
        private readonly CircleComponent _circle;

        private float _deathTimer = 1.0f;

        public Laser(Game game)
            : base(game)
        {
            // Create a sprite component
            var sprite = new SpriteComponent(this);
            sprite.SetTexture(Game.Renderer.GetTexture("Assets/Laser.png"));
            
            // Create a move component, and set a forward speed
            var move = new MoveComponent(this)
            {
                ForwardSpeed = 800.0f,
            };

            // Create a circle component (for collision)
            _circle = new CircleComponent(this)
            {
                Radius = 11.0f,
            };
        }

        protected override void UpdateActor(float deltaTime)
        {
            // If we run out of time, laser is dead
            _deathTimer -= deltaTime;
            if (_deathTimer <= 0.0f)
            {
                State = ActorState.Dead;
            }
            else
            {
                // Do we intersect with an asteroid?
                foreach (var asteroid in Game.Asteroids)
                {
                    // The first asteroid we intersect with,
                    // set ourselves and the asteroid to dead
                    if (CircleComponent.Intersect(_circle, asteroid.Circle))
                    {
                        State = ActorState.Dead;
                        asteroid.State = ActorState.Dead;
                        break;
                    }
                }
            }
        }
    }
}